source/Public/Revoke-SqlDscServerPermission.ps1

```diff
@@ -0,0 +1,136 @@
+
+    .PARAMETER WithGrant
+        Specifies that the right to grant the permission should also be revoked,
+        and the revocation will cascade to other principals that the grantee has
+        granted the same permission to.
+
+    .PARAMETER Force
+        Specifies that the permissions should be revoked without any confirmation.
+
+    .OUTPUTS
+        None.
+
+    .EXAMPLE
+        $serverInstance = Connect-SqlDscDatabaseEngine
+
+        $permissionSet = [Microsoft.SqlServer.Management.Smo.ServerPermissionSet] @{
+            ConnectSql = $true
+            ViewServerState = $true
+        }
+
+        Revoke-SqlDscServerPermission -ServerObject $serverInstance -Name 'MyPrincipal' -Permission $permissionSet
+
+        Revokes the specified permissions from the principal 'MyPrincipal'.
+
+    .EXAMPLE
+        $serverInstance = Connect-SqlDscDatabaseEngine
+
+        $permissionSet = [Microsoft.SqlServer.Management.Smo.ServerPermissionSet] @{
+            AlterAnyDatabase = $true
+        }
+
+        $serverInstance | Revoke-SqlDscServerPermission -Name 'MyPrincipal' -Permission $permissionSet -WithGrant -Force
+
+        Revokes the specified permissions and the right to grant them from the principal 'MyPrincipal' with cascading effect, without prompting for confirmation.
+
+        Revokes the specified permissions (including grant options) from the principal 'MyPrincipal' without prompting for confirmation.
+
+    .NOTES
+        If specifying `-ErrorAction 'SilentlyContinue'` then the command will silently
+        ignore if the principal is not present. If specifying `-ErrorAction 'Stop'` the
+        command will throw an error if the principal is missing.
+
+        When revoking permission with State 'GrantWithGrant', both the
+        grantee and all the other users the grantee has granted the same permission
+        to, will also get their permission revoked.
+#>
+function Revoke-SqlDscServerPermission
+{
+    [System.Diagnostics.CodeAnalysis.SuppressMessageAttribute('UseSyntacticallyCorrectExamples', '', Justification = 'Because the rule does not yet support parsing the code when a parameter type is not available. The ScriptAnalyzer rule UseSyntacticallyCorrectExamples will always error in the editor due to https://github.com/indented-automation/Indented.ScriptAnalyzerRules/issues/8.')]
+    [System.Diagnostics.CodeAnalysis.SuppressMessageAttribute('AvoidThrowOutsideOfTry', '', Justification = 'Because the code throws based on an prior expression')]
+    [CmdletBinding(SupportsShouldProcess = $true, ConfirmImpact = 'High')]
+    [OutputType()]
+    param
+    (
+        [Parameter(Mandatory = $true, ValueFromPipeline = $true)]
+        [Microsoft.SqlServer.Management.Smo.Server]
+        $ServerObject,
+
+        [Parameter(Mandatory = $true)]
+        [System.String]
+        $Name,
+
+        [Parameter(Mandatory = $true)]
+        [Microsoft.SqlServer.Management.Smo.ServerPermissionSet]
```

Comment:
I think we should have the type  [Microsoft.SqlServer.Management.Smo.ServerPermission] here instead so the user specifies for example `-Permission 'ViewServerState' ,'ConnectSql'`.

Thtoughout the relavant commands.